mkosi/util.py

```diff
@@ -77,8 +77,13 @@ def flatten(lists: Iterable[Iterable[T]]) -> list[T]:
 class INVOKING_USER:
     uid = int(os.getenv("SUDO_UID") or os.getenv("PKEXEC_UID") or os.getuid())
     gid = int(os.getenv("SUDO_GID") or os.getgid())
-    name = pwd.getpwuid(uid).pw_name
-    home = Path(f"~{name}").expanduser()
+    try:
+        name = pwd.getpwuid(uid).pw_name
+        home = Path(f"~{name}").expanduser()
+    except KeyError:
+        # names for UIDs might not be always resolvable, e.g. during reproducible builds
+        name = "fallback"
+        home = Path("/")
```

Comment:
How about adding more fallbacks before we have to resort to something hardcoded?
In the linked build log `SUDO_USER=jenkins` is shown to be available. In my local shell also `USER=`, `USERNAME=` and `LOGNAME` are available and seem like reasonable variables to check. Also checking `HOME=` would make sense. Home and logname are also part of POSIX.

Apart from that I think that this should show a warning at least. My worry is people with weird combinations running mkosi and not knowing that they triggered this fallback (e.g. some weird combination of systemd-homed or LDAP (which will not show up in the password database) and privilege escalations like doas which do not variables containing the original user.

Reply:
The `KeyError` will be raised if the numeric UID cannot be resolved to a username. As long as people have some form of working NSS that should work. That not being available should hopefully be a very niche problem, otherwise people will already have a problem. Adding `SUDO_USER`, `USER`, `USERNAME` and `LOGNAME`, would introduce edge cases where we might not be entirely sure anymore what we're getting, so I would prefer to have a clear failure mode. If this is triggered too easily, then we should revisit this.

Does this make sense?

Reply:
> Adding `SUDO_USER`, `USER`, `USERNAME` and `LOGNAME`, would introduce edge cases where we might not be entirely sure anymore what we're getting...

`SUDO_UID` and `PKEXEC_UID` are already used an I fail to see how they are different than `SUDO_USER` or `LOGNAME`. Also `PKEXEC_UID` has no gid equivalent so inconsistent states are already possible.

> ...so I would prefer to have a clear failure mode. If this is triggered too easily, then we should revisit this.

This is why I would definitely emit a warning whenever the fallback is triggered.

Reply:
Note that this is only a problem for the unit tests, so having a fallback that is only enabled when being run from a unit test would be perfectly fine for me

Reply:
I find the idea appealing to limit this just to tests, but since this explodes during import time this can't be mocked out easily. Is setting an environment variable for the test an option? In that case the try-except-block could just be a check whether the magic variable is set. This would make the whole behaviour opt in and sidestep all questions of people triggering this by accident.

Reply:
Yeah sure, that's not an issue

Reply:
Reworked like that. This also has the added benefit of making this behaviour easily testable.